### PR TITLE
make `startTaskPipeline` iterative

### PR DESCRIPTION
**Pull request by freben**
Originated from discussion in #17249 

Let's start small. This just replicates what we had. Then I can add additional logging and watchdog etc on top of this in a separate PR.

The old task pipeline tests are unchanged and green.

### CODE REVIEW

File: plugins/catalog-backend/src/processing/util.ts [hunk 1]
```diff
@@ -86,3 +86,50 @@ export function isObject(value: JsonValue | undefined): value is JsonObject {
...
+  }
+
+  function release() {
+    const resolversToCall = resolvers.splice(0, resolvers.length);
+    for (const resolver of resolversToCall) {
+      resolver();
+    }
+  }
+
+  signal.addEventListener('abort', release);
```
Comment (Rugvip) [inline plugins/catalog-backend/src/processing/util.ts:132]: Probably safest to remove this afterwards

File: plugins/catalog-backend/src/processing/util.ts [hunk 2]
```diff
@@ -86,3 +86,50 @@ export function isObject(value: JsonValue | undefined): value is JsonObject {
...
+ */
+export function createBarrier(options: {
+  waitTimeoutMillis: number;
+  signal: AbortSignal;
+}): {
+  wait: () => Promise<void>;
+  release: () => void;
+} {
+  const { waitTimeoutMillis, signal } = options;
+  const resolvers: Array<() => void> = [];
```
Comment (Rugvip) [inline plugins/catalog-backend/src/processing/util.ts:102]: Bit simpler to make this a `Set` I'd think

File: plugins/catalog-backend/src/processing/TaskPipeline.ts [hunk 3]
```diff
@@ -72,50 +74,58 @@ export function startTaskPipeline<T>(options: Options<T>) {
...
-  // dry because of the lack of available tasks. As long as there are
-  // enough items to process this will be a noop.
-  const intervalId = setInterval(() => {
-    maybeLoadMore();
-  }, pollingIntervalMs);
+  pipelineLoop().catch(_error => {
+    // This should be impossible, but if it did happen, it would signal a
+    // programming error inside the loop (errors should definitely be caught
+    // inside of it).
+  });
```
Comment (Rugvip) [inline plugins/catalog-backend/src/processing/TaskPipeline.ts:126]: We don't want this to fail silently do we though? Probably best to re-throw as an uncaught exception to crash the backend?

File: plugins/catalog-backend/src/processing/util.ts [hunk 4]
```diff
@@ -86,3 +86,50 @@ export function isObject(value: JsonValue | undefined): value is JsonObject {
 export const validateEntity = entitySchemaValidator();
 
 export const validateEntityEnvelope = entityEnvelopeSchemaValidator();
+
+/**
+ * Creates a barrier with a timeout, that can be awaited or prematurely
+ * released either manually or by an abort signal.
+ */
+export function createBarrier(options: {
```
Comment (Rugvip) [inline plugins/catalog-backend/src/processing/util.ts:94]: This feels like a very specific barrier implementation that you wouldn't normally expect. Especially the behavior of not clearing the resolvers on timeout. Maybe best to just put this next to the `TaskPipeline` implementation instead? in the same file I mean

**Review by Rugvip — APPROVED** [review]
Sweet 👌

File: plugins/catalog-backend/src/processing/util.ts (hunk 1 above)
Comment (freben) [reply plugins/catalog-backend/src/processing/util.ts:132]: There is no real afterwards though.

Comment (freben) [reply plugins/catalog-backend/src/processing/util.ts:132]: Moved this into a dedicated destroy function that gets called when the pipeline is stopped, for completeness

File: plugins/catalog-backend/src/processing/util.ts (hunk 2 above)
Comment (freben) [reply plugins/catalog-backend/src/processing/util.ts:102]: Agreed! Did that now

File: plugins/catalog-backend/src/processing/TaskPipeline.ts (hunk 3 above)
Comment (freben) [reply plugins/catalog-backend/src/processing/TaskPipeline.ts:126]: We don't crash on those anymore do we? 🤔

File: plugins/catalog-backend/src/processing/util.ts (hunk 4 above)
Comment (freben) [reply plugins/catalog-backend/src/processing/util.ts:94]: Aye, moved

File: plugins/catalog-backend/src/processing/TaskPipeline.ts (hunk 3 above)
Comment (freben) [reply plugins/catalog-backend/src/processing/TaskPipeline.ts]: Alright, updated this with a rethrow, it'll make it log which is excellent either way.

File: plugins/catalog-backend/src/processing/TaskPipeline.ts [hunk 5]
```diff
@@ -72,50 +72,112 @@ export function startTaskPipeline<T>(options: Options<T>) {
...
     }
   }
 
-  // This interval makes sure that we load in new items if the loop runs
-  // dry because of the lack of available tasks. As long as there are
-  // enough items to process this will be a noop.
-  const intervalId = setInterval(() => {
-    maybeLoadMore();
-  }, pollingIntervalMs);
+  signal.addEventListener('abort', release);
```
Comment (Rugvip) [inline plugins/catalog-backend/src/processing/TaskPipeline.ts:176]: hodl, just removing it in the callback would be enough right? As in

```ts
  function release() {
    const resolversToCall = new Set(resolvers);
    resolvers.clear();
    for (const resolver of resolversToCall) {
      resolver();
    }
    signal.removeEventListener('abort', release)
  }
```

Comment (freben) [reply plugins/catalog-backend/src/processing/TaskPipeline.ts:176]: No, release is called over and over again. It'd render this subscription immediately useless after the first task completes. Note that this addition is made only exactly once.

Comment (Rugvip) [reply plugins/catalog-backend/src/processing/TaskPipeline.ts:176]: Well with a different function for the signal abort callback

```ts
function onAbort() {
  release()
  signal.removeEventListener('abort', onAbort)
}
signal.addEventListener('abort', onAbort);
```

Comment (freben) [reply plugins/catalog-backend/src/processing/TaskPipeline.ts:176]: Mm. I wonder how much we need to worry though. It uses a weak map internally afaik so it should not leak

**Pull request merged**